Created a single class business layer

paymog opened the pull request:
Before we go further into this, can you guys critique the fuck out of what I'm doing?

My next plan is to move some/most/all of the logic from CryptogramSolveWidget to CrytpogramManager.

I envision that each UI component will have some sort of xxxManager class to handle the logic it needs

paymog committed:
create branch and first BL object

Created a cryptogram manager for the business layer

paymog committed:
Put CrytpgramManager into CryptogramEditWidget

I removed the logic from CryptogramEditWidget

All logic should be in CryptogramManager now

rikimbo commented on pg13/business/CryptogramManager.java:
```diff
@@ -0,0 +1,48 @@
+			if ("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890.,:;- !?()&#$%'\"".indexOf(curr) < 0)
+			{
+				throw new IllegalArgumentException("Invalid character in plaintext");
+			}
```
Should we really be throwing an exception for this?  This function is used by the GUI to validate text entered, and the user is going to try to type in all sorts of garbage.  Is it really an exceptional case?  Maybe it could be restructured to return a boolean.

EDIT:  Actually maybe it's legit to do it like that.  You've done a lot more GUI interaction then I have.  Is that what people do?

paymog replied:
Well, in WPF it's not done like that at all. However, WPF is an entirely different environment. Let me read my exceptions guideline and see if I can come up with a more definitive answer :p

rikimbo commented on pg13/presentation/CryptogramEditWidget.java:
```diff
@@ -9,101 +9,73 @@
-	            	case SWT.ARROW_RIGHT:  // Right arrow  
-	                textIsValid = true;  
-				}  
-				
```
You shouldn't remove the part of the validation that checks the keycode for backspace, delete, and other control keys.  It might prevent the user from doing those things in the text box.  Not sure though.  Did you test it?

paymog replied:
Yup, all of those work. I totally forgot to test for that though.

rikimbo replied:
Oh okay.  I looked up online how to write one of those listenners, and it told me to include that, so that's why it's there.  If it's not necessary, it's fine to take it out.

paymog replied:
TBH, I'm not sure. I almost literally just copy-pasted the code from the GUI file to the business file.

rikimbo replied:
Yeah, you do all the checking of the text that I was doing, just not the checking of the keycode.  When I put it in, it didn't feel like it should be there because I thought the verify listenner was just going to verify the text entered, and not mess around with keycodes in the first place.  Again, I just put the keycode checking because it  was in there on the solution I found for that on the internet.  I'm not on my home machine, otherwise I'd test it in a windows environment to make sure it works for that too.

rikimbo commented on pg13/presentation/CryptogramEditWidget.java:
```diff
@@ -9,101 +9,73 @@
+			public void verifyText(VerifyEvent event) {
+				try {
+					cm.validatePlaintext(event.text);
+				} catch (IllegalArgumentException e) {
```
I like the call to validate to a CM instead of doing this directly in here.  Still unsure about using an exception for it though.  See my other comment.

paymog replied:
I'm also unsure of whether an exception is the best way to handle it. I debated a boolean but my gut told me exception is the way to go. I'm not sure if I can defend my choice very well though.

rikimbo replied:
I just feel funny about the possibility of throwing an exception every time the listenner catches a character and passes it to validate.  The user will never see them, but the fact that we expect it to happen so commonly here makes me a little bit uneasy about it.

paymog replied:
I agree. Honestly, I'm not sure if us preventing them from entering a character is a good idea. Purely from an HCI perspective, if I type a character, I expect that character to become visible on the screen. Maybe it's better to let them enter anything they want and then tell them they have an invalid phrase when the press the generate preview? Thoughts?

rikimbo replied:
Well, the reason I did it that was was because for a field where you're expected to enter letters or numbers (ie for a product key in an installer, for example), all it does it limit your input that way too.  For the plaintext window, really all it disallows are accented letters, and non-punctuation symbols like @ or ~ that shouldn't show up in text anyway.  We could do the validation at the point when we hit preview, but my feeling with that kind of thing has always been "Grr! Then why did you let me enter it in the first place?".  That's my preference, though.  Really it's just tomato tomahto.

paymog replied:
Good point. It doesn't matter for me either way. However, the question of exception vs boolean is still on the table.

rikimbo replied:
I'm actually warming up to it, since I like the way it makes the validation functions look.  I could go either way on this.

rikimbo commented on pg13/presentation/CryptogramEditWidget.java:
```diff
@@ -111,13 +83,11 @@ public void verifyText(VerifyEvent event)
+
 		Button btnPreview = new Button(this, SWT.NONE);
-		btnPreview.addSelectionListener(new SelectionAdapter() 
-		{
```
Why did you change the bracketing?  I thought our standards said the bracketing should be on a separate line.

paymog replied:
Totally my bad. I must have done used the autoformat funciton of eclipse. I'll change that ASAP

paymog replied:
K, I'm gonna make an eclipse formatting thing. I'm looking at the coding standards and don't see anything about braces on the different lines. Should I add that to the coding standards?

rikimbo replied:
There's a few other places where changes are marked for bracket changes, this is the only one I commented on though.

rikimbo replied:
Oh, I thought it was one of the things that was brought up when we were talking about standards.  Sure, add it to the document, I think we've been doing that so far.  Well, speaking for myself, I guess.

rikimbo commented on pg13/presentation/CryptogramEditWidget.java:
```diff
@@ -131,20 +101,25 @@ public void widgetSelected(SelectionEvent e)
+		} catch (IllegalArgumentException e) {
+			MessageBox dialog = new MessageBox(this.getShell(), SWT.OK);
+			dialog.setText("Invalid text");
+			dialog.setMessage("The text you have entered is invalid. Please modify the text.");
```
This shouldn't ever happen, since everything that's entered into the plaintext box is intercepted by a listenner there, but still, it's good to have it.

EDIT:  I like the use of an exception here.  Because this is something that really shouldn't happen, when it does manage to get through, we catch it and throw a prompt.  This one I am happy with.

rikimbo commented:
Regarding the UI elements having an xxxManager for all the components it needs:

Most of the UI formatting is handled by SWT and the Layout objects.  The only component where that is done manually is in the SolveWidget (or whatever I called it) to organize and display the LetterWidgets.  That part interacts directly with LayoutData objects and Widgets and stuff, so as much as I'd like to see that process move to the business layer, I think it might be unfeasable or end up looking really messy.

If you find any other logic in the presentation layer that you can move to the business layer or make a manager for, go ahead and give it a shot.  I like that you've moved the text validation to the business layer.  Check out the LetterWidget, there's validation for the text field there too, but it's more restrictive.  What's entered there needs to be a lowercase or uppercase letter (isLetter won't work since it accept accents and stuff).  Additionally, that listenner intercepts the event and sets the text to lowercase.

#############
I'm not sure how to reply to this comment so I'll just extend it.

I don't think it makes sense to move layout logic to the business layer. Since that logic is highly coupled (right word) to the UI. I was thinking of moving other logic, similar to the validation, to the business layer. I'd like to hear from Will and Lauren too just to make sure everyone is on board with my interpretation of how the business layer should look (from a code perspective) and behave.

#############
(Eric speaking again)
I agree with what you said.  Let's leave the logic involved in the arrangement and layout of gui elements with the presentation layer, but move any validation of input or objects to the business layer.

paymog commented on tests/business/CryptogramManagerTest.java:
```diff
@@ -0,0 +1,89 @@
+	 * @author PaymahnMoghadasian
+	 * @date June 1 2013
+	 * 
+	 * I'm not quite sure what constitutes invalid plaintext for a cryptogram. Somemone should extend these tests.
```
Someone who knows what constitutes an invalid cryptogram should either fill in this method or give me examples so I can do it

rikimbo replied:
Just throw in accented letters, and other non-letters we didn't allow, eg *, @, =, whatever.  That really is all we're disallowing in the plaintext, I think.

paymog replied:
Will do

rikimbo commented:
Why did Travis fail on this?  I thought we had the little guy working.

rikimbo commented:
I think we're on the right track doing what you're doing here.  Do you want this merged now, or do you want to make some changes and keep committing?  Will will have some requirements from the business layer too, when he starts working on implementing the Find Screen, probably.

Same deal for when the Save button is implemented.

############
(Paymahn speaking (writing?))
This can be merged right now. I'll create another branch to extend the functionality

rikimbo commented on pg13/presentation/CreateScreen.java:
```diff
@@ -47,6 +48,7 @@ public CreateScreen(Composite parent, int style)
 		fd_composite.top = new FormAttachment(0);
 		fd_composite.left = new FormAttachment(33);
 		cmpEditWidget.setLayoutData(fd_composite);
+		setTabList(new Control[]{cmpProperties, cmpEditWidget});
```
Does this fix the tabbing order?

paymog committed:
Added more functionality to Cryptogram manager

Cryptogram manager now handles more logic. I'm gonna try to integrate
it next.

BTW: I exported an eclipse autoformat thing which can be imported so we
all have the same formatting

paymog committed:
Starting integrating CryptogramManager into CryptogramLetterWidget

paymog committed:
Moved more logic from CryptogramLetterWidget to CryptogramManager

paymog commented:
Don't merge this. I've made more commits and I'm not sure what would happen if this gets merged and then I sync my commits.

paymog committed:
Fixed shit

Did some final changes. It seems that everything works.

You guys should give this a test run before you merge.

paymog committed:
Clean up
